File: dice.cpp
// dice.cpp
#include <iostream>
#include <vector>
#include <string>
#include <random>
#include <algorithm>
#include <numeric>
#include <iomanip>
#include <fstream>
#include <cctype>

using namespace std;

const string RESET = "\033[0m";
const string RED = "\033[91m";
const string GREEN = "\033[92m";
const string YELLOW = "\033[93m";
const string BLUE = "\033[94m";
const string MAGENTA = "\033[95m";
const string CYAN = "\033[96m";
const string BOLD = "\033[1m";

string colorize(const string& text, const string& color) {
    return color + text + RESET;
}

vector<string> DICE_COLORS = {RED, GREEN, YELLOW, BLUE, MAGENTA, CYAN};

vector<int> rollDice(int num, int faces, mt19937& rng) {
    vector<int> res;
    uniform_int_distribution<int> dist(1, faces);
    for (int i=0; i<num; ++i) res.push_back(dist(rng));
    return res;
}

string formatDice(const vector<int>& values, int faces) {
    string s;
    for (size_t i=0; i<values.size(); ++i) {
        int v = values[i];
        string col = DICE_COLORS[i % DICE_COLORS.size()];
        if (v == 1 || v == faces) col = BOLD;
        s += colorize(to_string(v), col) + " ";
    }
    return s;
}

int main(int argc, char* argv[]) {
    int numDice = 1, numFaces = 6, rolls = 1;
    bool showSum = false, showStats = false, verbose = false;
    string outputFile;
    random_device rd;
    mt19937 rng(rd());

    for (int i=1; i<argc; ++i) {
        string arg = argv[i];
        if (arg == "-h" || arg == "--help") {
            cout << "Usage: dice [num_dice] [num_faces] [-r rolls] [-s] [-t] [-v] [-o file]" << endl;
            return 0;
        } else if (arg == "-r" && i+1 < argc) {
            rolls = stoi(argv[++i]);
        } else if (arg == "-s") {
            showSum = true;
        } else if (arg == "-t") {
            showStats = true;
        } else if (arg == "-v") {
            verbose = true;
        } else if (arg == "-o" && i+1 < argc) {
            outputFile = argv[++i];
        } else if (numDice == 1 && arg.find_first_not_of("0123456789") == string::npos) {
            numDice = stoi(arg);
        } else if (numFaces == 6 && arg.find_first_not_of("0123456789") == string::npos) {
            numFaces = stoi(arg);
        }
    }
    if (numDice < 1 || numFaces < 2) {
        cerr << "Количество кубиков >= 1, граней >= 2" << endl;
        return 1;
    }

    vector<vector<int>> allResults;
    for (int i=0; i<rolls; ++i) {
        allResults.push_back(rollDice(numDice, numFaces, rng));
    }

    vector<string> lines;
    if (verbose) {
        for (size_t i=0; i<allResults.size(); ++i) {
            int sum = accumulate(allResults[i].begin(), allResults[i].end(), 0);
            string diceStr = formatDice(allResults[i], numFaces);
            lines.push_back("Бросок " + to_string(i+1) + ": " + diceStr + " (сумма: " + to_string(sum) + ")");
        }
    } else if (showSum) {
        for (auto& v : allResults) {
            lines.push_back(to_string(accumulate(v.begin(), v.end(), 0)));
        }
    } else {
        for (auto& v : allResults) {
            string line;
            for (int x : v) line += to_string(x) + " ";
            lines.push_back(line);
        }
    }

    if (showStats && rolls > 1) {
        vector<int> sums;
        for (auto& v : allResults) sums.push_back(accumulate(v.begin(), v.end(), 0));
        int mn = *min_element(sums.begin(), sums.end());
        int mx = *max_element(sums.begin(), sums.end());
        double avg = accumulate(sums.begin(), sums.end(), 0.0) / sums.size();
        sort(sums.begin(), sums.end());
        double med = (sums.size() % 2 == 0) ? (sums[sums.size()/2-1] + sums[sums.size()/2]) / 2.0 : sums[sums.size()/2];
        lines.push_back("\nСтатистика по суммам:");
        lines.push_back("  Минимум: " + to_string(mn));
        lines.push_back("  Максимум: " + to_string(mx));
        lines.push_back("  Среднее: " + to_string(avg));
        lines.push_back("  Медиана: " + to_string(med));
    }

    string output;
    for (auto& line : lines) output += line + "\n";
    if (!outputFile.empty()) {
        ofstream f(outputFile);
        if (f) { f << output; cout << colorize("Результат сохранён в " + outputFile, GREEN) << endl; }
        else cerr << colorize("Ошибка записи файла", RED) << endl;
    } else {
        cout << output;
    }
    return 0;
}

File: dice.go
// dice.go
package main

import (
	"bufio"
	"fmt"
	"math/rand"
	"os"
	"strconv"
	"strings"
	"time"
)

const (
	reset  = "\033[0m"
	red    = "\033[91m"
	green  = "\033[92m"
	yellow = "\033[93m"
	blue   = "\033[94m"
	magenta= "\033[95m"
	cyan   = "\033[96m"
	bold   = "\033[1m"
)

func colorize(text, color string) string {
	return color + text + reset
}

var diceColors = []string{red, green, yellow, blue, magenta, cyan}

func rollDice(num, faces int) []int {
	res := make([]int, num)
	for i := 0; i < num; i++ {
		res[i] = rand.Intn(faces) + 1
	}
	return res
}

func formatDice(values []int, faces int) string {
	var parts []string
	for i, v := range values {
		col := diceColors[i%len(diceColors)]
		if v == 1 || v == faces {
			col = bold
		}
		parts = append(parts, colorize(strconv.Itoa(v), col))
	}
	return strings.Join(parts, " ")
}

func main() {
	rand.Seed(time.Now().UnixNano())
	numDice := 1
	numFaces := 6
	rolls := 1
	showSum := false
	showStats := false
	verbose := false
	outputFile := ""

	args := os.Args[1:]
	for i := 0; i < len(args); i++ {
		arg := args[i]
		switch arg {
		case "-h", "--help":
			fmt.Println("Usage: dice [num_dice] [num_faces] [-r rolls] [-s] [-t] [-v] [-o file]")
			return
		case "-r":
			if i+1 < len(args) {
				rolls, _ = strconv.Atoi(args[i+1])
				i++
			}
		case "-s":
			showSum = true
		case "-t":
			showStats = true
		case "-v":
			verbose = true
		case "-o":
			if i+1 < len(args) {
				outputFile = args[i+1]
				i++
			}
		default:
			if numDice == 1 {
				if v, err := strconv.Atoi(arg); err == nil && v > 0 {
					numDice = v
				}
			} else if numFaces == 6 {
				if v, err := strconv.Atoi(arg); err == nil && v > 1 {
					numFaces = v
				}
			}
		}
	}
	if numDice < 1 || numFaces < 2 {
		fmt.Println("Количество кубиков >= 1, граней >= 2")
		return
	}

	var allResults [][]int
	for i := 0; i < rolls; i++ {
		allResults = append(allResults, rollDice(numDice, numFaces))
	}

	var lines []string
	if verbose {
		for i, v := range allResults {
			sum := 0
			for _, x := range v {
				sum += x
			}
			diceStr := formatDice(v, numFaces)
			lines = append(lines, fmt.Sprintf("Бросок %2d: %s  (сумма: %d)", i+1, diceStr, sum))
		}
	} else if showSum {
		for _, v := range allResults {
			sum := 0
			for _, x := range v {
				sum += x
			}
			lines = append(lines, strconv.Itoa(sum))
		}
	} else {
		for _, v := range allResults {
			parts := make([]string, len(v))
			for i, x := range v {
				parts[i] = strconv.Itoa(x)
			}
			lines = append(lines, strings.Join(parts, " "))
		}
	}

	if showStats && rolls > 1 {
		sums := make([]int, len(allResults))
		for i, v := range allResults {
			s := 0
			for _, x := range v {
				s += x
			}
			sums[i] = s
		}
		mn, mx := sums[0], sums[0]
		sumAll := 0
		for _, s := range sums {
			if s < mn {
				mn = s
			}
			if s > mx {
				mx = s
			}
			sumAll += s
		}
		avg := float64(sumAll) / float64(len(sums))
		// median
		sorted := make([]int, len(sums))
		copy(sorted, sums)
		sort.Ints(sorted)
		var med float64
		if len(sorted)%2 == 0 {
			med = float64(sorted[len(sorted)/2-1]+sorted[len(sorted)/2]) / 2.0
		} else {
			med = float64(sorted[len(sorted)/2])
		}
		lines = append(lines, "\nСтатистика по суммам:")
		lines = append(lines, fmt.Sprintf("  Минимум: %d", mn))
		lines = append(lines, fmt.Sprintf("  Максимум: %d", mx))
		lines = append(lines, fmt.Sprintf("  Среднее: %.2f", avg))
		lines = append(lines, fmt.Sprintf("  Медиана: %.2f", med))
	}

	output := strings.Join(lines, "\n")
	if outputFile != "" {
		err := os.WriteFile(outputFile, []byte(output), 0644)
		if err == nil {
			fmt.Println(colorize("Результат сохранён в "+outputFile, green))
		} else {
			fmt.Println(colorize("Ошибка записи файла", red))
		}
	} else {
		fmt.Println(output)
	}
}
